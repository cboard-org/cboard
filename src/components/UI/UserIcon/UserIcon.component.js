import React from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import { withStyles } from '@material-ui/core/styles';
import { Link } from 'react-router-dom';
import Avatar from '@material-ui/core/Avatar';
import green from '@material-ui/core/colors/green';

import AccountIcon from '@material-ui/icons/AccountCircle';

import IconButton from '../IconButton';
import messages from './UserIcon.messages';

const propTypes = {
  /**
   * @ignore
   */
  intl: intlShape.isRequired,
  user: PropTypes.object,
  link: PropTypes.bool,
  accountIcon: PropTypes.any
};

const styles = {
  greenAvatar: {
    backgroundColor: green[500]
  }
};

export const UserIcon = ({
  intl,
  user,
  classes,
  link = true,
  accountIcon = AccountIcon,
  ...other
}) => {
  let avatar = null;
  let hasPhotos = false;
  if (user) {
    const fPhotos =
      user.facebook && user.facebook.photos && user.facebook.photos.length
        ? user.facebook.photos
        : null;
    const gPhotos =
      user.google && user.google.photos && user.google.photos.length
        ? user.google.photos
        : null;
    const uPhotos = user.photos && user.photos.length ? user.photos : null;
    const photos = uPhotos || fPhotos || gPhotos || [];

    if (photos.length) {
      hasPhotos = true;
      avatar = photos[0];
    } else {
      const [first, second = ''] = user.name
        .toUpperCase()
        .split(' ')
        .slice(0, 2);
      avatar = `${first.slice(0, 1)}${second.slice(0, 1)}`;
    }
  }

  const AccountIconToRender = accountIcon;

  return (
    <IconButton
      label={user ? user.name : intl.formatMessage(messages.login)} {...other}
    >
      <React.Fragment>
        {!user && link && (
          <Link to="/login-signup" className="LoginSignUpButton">
            <AccountIconToRender />
          </Link>
        )}
        {!user && !link && (
          <Avatar>
            <AccountIconToRender />
          </Avatar>
        )}
        {!!user && (
          <Avatar className={`UserIcon__Avatar ${classes.greenAvatar}`}>
            {!hasPhotos && avatar}
            {hasPhotos && <img src={avatar} alt={user.name} />}
          </Avatar>
        )}
      </React.Fragment>
    </IconButton>
  );
};

UserIcon.propTypes = propTypes;

export default withStyles(styles)(injectIntl(UserIcon));
