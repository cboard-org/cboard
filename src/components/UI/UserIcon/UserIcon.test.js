import React from 'react';
import { shallowMatchSnapshot } from '../../../common/test_utils';
import { UserIcon } from './UserIcon.component';

jest.mock('./UserIcon.messages', () => {
  return {
    login: {
      id: 'cboard.components.UserIcon.login',
      defaultMessage: 'Login or Sign up'
    }
  };
});

const COMPONENT_PROPS = {
  user: null,
  intl: {
    formatMessage: () => 'label'
  },
  classes: {
    greenAvatar: 'some-green-class'
  }
};

describe('UserIcon tests', () => {
  test('default renderer', () => {
    shallowMatchSnapshot(<UserIcon {...COMPONENT_PROPS} />);
  });

  test('with user', () => {
    const props = {
      ...COMPONENT_PROPS,
      user: {
        name: 'Some Name'
      }
    };
    shallowMatchSnapshot(<UserIcon {...props} />);
  });

  test('with user with facebook photos', () => {
    const props = {
      ...COMPONENT_PROPS,
      user: {
        name: 'Some Name',
        facebook: {
          photos: ['some-picture-link.png']
        }
      }
    };
    shallowMatchSnapshot(<UserIcon {...props} />);
  });
});
