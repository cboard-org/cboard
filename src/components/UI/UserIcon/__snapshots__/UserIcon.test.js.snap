// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserIcon tests default renderer 1`] = `
<IconButton
  label="label"
>
  <React.Fragment>
    <Link
      className="LoginSignUpButton"
      replace={false}
      to="/login-signup"
    >
      <pure(AccountCircleIcon) />
    </Link>
  </React.Fragment>
</IconButton>
`;

exports[`UserIcon tests with user 1`] = `
<IconButton
  label="Some Name"
>
  <React.Fragment>
    <WithStyles(Avatar)
      className="UserIcon__Avatar some-green-class"
    >
      SN
    </WithStyles(Avatar)>
  </React.Fragment>
</IconButton>
`;

exports[`UserIcon tests with user with facebook photos 1`] = `
<IconButton
  label="Some Name"
>
  <React.Fragment>
    <WithStyles(Avatar)
      className="UserIcon__Avatar some-green-class"
    >
      <img
        alt="Some Name"
        src="some-picture-link.png"
      />
    </WithStyles(Avatar)>
  </React.Fragment>
</IconButton>
`;
